File: src/server/src/test/test_velvet_dawn/test_datapacks/test_units/test_unit_upgrade_parsing.py
import velvet_dawn.datapacks
from test.base_test import BaseTest
from velvet_dawn import errors
from velvet_dawn.models import Attributes
import velvet_dawn


class TestUnitUpgradesParsing(BaseTest):

    def test_upgrade_parsing(self):
        # Test no name
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {})
        velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {"name": "hi"})

        # Type id is wrong type
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {
                "id": False,
                "name": "hi"
            })

        # Enabled is not a list
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {
                "enabled": False,
                "name": "hi"
            })

        # Hidden is not a list
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {
                "hidden": False,
                "name": "hi"
            })

        # actions is not a list
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {
                "actions": False,
                "name": "hi"
            })

        # Require is not a list of strings
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {
                "requires": False,
                "name": "hi"
            })
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {
                "requires": [False],
                "name": "hi"
            })

        # Random key
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {
                "random-key": [False],
                "name": "hi"
            })

        # Valid
        velvet_dawn.models.datapacks.units.upgrades.Upgrade.load("0", -1, {
            "id": "example",
            "enabled": [{"if": "self", "eq": "example"}],
            "hidden": [{"if": "self", "ne": "example"}],
            "actions": [{"modify": "self.health", "add": 1}],
            "requires": ["eg"],
            "name": "hi"
        })

File: src/jserver/src/test/velvetdawn/datapacks/entities/TestUnitUpgradeParsing.java
package test.velvetdawn.datapacks.entities;

import org.junit.Test;
import test.BaseTest;
import velvetdawn.mechanics.upgrades.Upgrades;
import velvetdawn.models.anytype.AnyJson;
import velvetdawn.models.anytype.AnyList;

import java.util.List;

import static org.junit.Assert.assertThrows;


public class TestUnitUpgradeParsing extends BaseTest {

    @Test
    public void test_upgrade_parsing() throws Exception {
        var velvetDawn = this.prepareGame();

        // Test no name
        assertThrows(Exception.class, () ->
                new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson())));
        new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson().set("name", "hi")));

        // Type id is wrong type
        assertThrows(Exception.class, () ->
            new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson()
                    .set("id", false)
                    .set("name", "hi"))));

        // Enabled is not a list
        assertThrows(Exception.class, () ->
            new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson()
                    .set("enabled", false).set("name", "hi"))));

        // Hidden is not a list
        assertThrows(Exception.class, () ->
            new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson()
                    .set("hidden", false)
                    .set("name", "hi"))));

        // actions is not a list
        assertThrows(Exception.class, () ->
            new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson()
                    .set("actions", false).set("name", "hu"))));

        // Require is not a list of strings
        assertThrows(Exception.class, () ->
            new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson()
                    .set("requires", false).set("name", "hi"))));
        assertThrows(Exception.class, () ->
            new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson()
                    .set("requires", AnyList.of(new AnyJson()))
                    .set("name", "hi"))));

        // Random key
        assertThrows(Exception.class, () ->
            new Upgrades().load(velvetDawn, "", AnyList.of(new AnyJson()
                    .set("random-key", AnyList.of("False"))
                    .set("name", "hi"))));

        // Valid
        new Upgrades().load(velvetDawn, "0", AnyList.of(new AnyJson()
                .set("id", "example")
                .set("enabled", AnyList.of(new AnyJson().set("if", "self").set("eq", "example")))
                .set("hidden", AnyList.of(new AnyJson().set("if", "self").set("ne", "example")))
                .set("actions", AnyList.of(new AnyJson().set("modify", "self.health").set("add", 1)))
                .set("requires", AnyList.of("eg"))
                .set("name", "hi")));
    }
}
